src/Common/src/Interop/Windows/kernel32/Interop.GetConsoleTitle.cs

```diff
@@ -20,53 +21,69 @@ internal partial class Kernel32
         {
             int lastError = 0;
-            StringBuilder sb = new StringBuilder(CharCountToByteCount(InitialBufferSizeInChars + 1));
-            int len = GetConsoleTitle(sb, sb.Capacity);
 
-            if (len <= 0)
-            {
-                lastError = Marshal.GetLastWin32Error();
-            }
-            else if (len > MaxAllowedBufferSizeInChars)
-            {
-                // Title is greater than the allowed buffer so we do not read the title and only pass the length to the caller.
-                title = string.Empty;
-                titleLength = len;
-                return 0;
-            }
-            else
+            // See note below for why we're pessimistically setting the capacity as if it was bytes, even though it's in chars.
+            StringBuilder sb = new StringBuilder(InitialBufferSizeInChars * sizeof(char));
+
+            while (true)
             {
-                if (len > InitialBufferSizeInChars)
+                // If capacity is insufficient, sometimes this function returns length,
+                // and sometimes (Windows 10 RS2) it returns 0 with either ERROR_INSUFFICIENT_BUFFER or ERROR_SUCCESS.
+                // In either of those latter cases, try again hopefully with a larger buffer.
+                //
+                // In some cases (Windows 7, perhaps depending on codepage) it interprets the second parameter as bytes. 
+                // Give it characters, pssimistically, but make sure that the number of characters we offer is twice what it claims it needs.
+                //
+                // In some cases (Windows 10 RS2), the returned length includes null, in others (Windows 7) it does not. 
+                // Pessimistically assume it does not.
+                int len = GetConsoleTitle(sb, sb.Capacity);
+
+                if (len <= 0)
                 {
-                    // We need to increase the sb capacity and retry.
-                    sb.Capacity = CharCountToByteCount(len + 1);
-                    len = GetConsoleTitle(sb, sb.Capacity);
-                    if (len <= 0)
+                    lastError = Marshal.GetLastWin32Error();
+
+                    if (len < 0 || (lastError != Errors.ERROR_INSUFFICIENT_BUFFER && lastError != Errors.ERROR_SUCCESS))
                     {
-                        lastError = Marshal.GetLastWin32Error();
+                        title = string.Empty;
+                        titleLength = title.Length;
+                        return lastError;
                     }
                 }
-            }
+                else if (sb.Capacity >= (len + 1) * sizeof(char))
+                {
+                    // Success
+                    title = sb.ToString();
+                    titleLength = title.Length;
+                    return 0;
+                }
+
+                // We need to increase the sb capacity and retry.
+                if (sb.Capacity >= MaxAllowedBufferSizeInChars * sizeof(char))
```

Comment:
What point is there in checking for a max size? Shouldn't failing normally be good enough?